### DISCUSSION TITLE
🎉  PaddleSpeech 实现多种卡通音色和方言的中英文混合 TTS

### DISCUSSION THREAD

**Discussion by yazone** [PaddlePaddle/PaddleSpeech#2492] [Show and tell]
实现单模型TTS中英文发音初步尝试了3种方案：单发音人方式、多发音人语料混合方式、finetune方式。

示例句子：
```text
大家好，我是parrot虚拟老师，我们来读一首诗，我与春风皆过客， I and the spring breeze are passing by,你携秋水揽星河， you take the autumn water to take the galaxy。
```

## （一）单发音人方式（有数据才行！）
使用同一发音人中文、英文语料，这种方式效果最佳，听听效果：

中英文女：

https://user-images.githubusercontent.com/10195479/193219058-5a98685b-c93d-4c74-8356-914c781b4d27.mp4

中英文男：

https://user-images.githubusercontent.com/10195479/193215189-501ab75b-7eca-4675-a304-2ad7488922a5.mp4

## （二）多发音人语料混合方式（音色串了！）
有开源数据中文baker、英文ljspeech数据，将数据混合成单发音人训练出一个模型，能听出一句话中文是baker、英文是ljspecch两种音色，我想音色差别不大的时候可以选用这种方式，听听效果：

https://user-images.githubusercontent.com/10195479/193216645-30394fd8-d4e4-4e4b-9612-436f130cfad9.mp4

## （三）finetune方式（音色一致了！）
在第一种中英文预训练模型上进行finetune，实现特色语音、动漫语音、方言，效果如下：

### 特色普通话
Baker说中英文：

https://user-images.githubusercontent.com/10195479/193209914-fdc5e771-c7bc-4be3-b6ef-00016f0089a5.mp4

抖音鸡汤女说中英文：

https://user-images.githubusercontent.com/10195479/193210117-774d58e4-eefd-4c8a-8018-826fc2b1d4ae.mp4

### 动漫

蜡笔小新说中英文：

https://user-images.githubusercontent.com/10195479/193217824-2eda00ba-1760-4da9-adf9-91beae36cd36.mp4

海绵宝宝说中英文：

https://user-images.githubusercontent.com/10195479/193218064-f0111d64-7d6d-4e2a-918e-58ebe228b9cb.mp4

### 方言
东北话：

https://user-images.githubusercontent.com/10195479/193218466-4213202a-b3ac-4ce6-9d8f-cd09e641c155.mp4

广西话：

https://user-images.githubusercontent.com/10195479/193218528-f191e296-fb64-4cfc-90a0-22407c04eb2c.mp4

河南话：

https://user-images.githubusercontent.com/10195479/193218587-0a5cd848-c6ae-44e5-956f-980e9f233bd7.mp4

四川话：

https://user-images.githubusercontent.com/10195479/193218647-84c18656-de58-499e-a180-67b2aae1b9b3.mp4

天津话：

https://user-images.githubusercontent.com/10195479/193218707-944f5801-2894-4e64-83c2-c92176df63cb.mp4

粤语：

https://user-images.githubusercontent.com/24568452/193232486-25b64f7a-164a-4433-8e7e-02f11c4036d4.mp4

<br>

后续是否可以在第二种方案的模型上进行finetune达到好的效果再进行验证，这样就不需要单发音人的中英文预训练模型了，只需要开源数据就OK了。

**×××××××××××PaddleTTS实现起来还是很方便的，感谢各位大佬...××××××××××**

**Comment by yt605155624** [comment]
`（二）多发音人语料混合方式（音色串了！）`

我们目前的中英文混合是用 4 个开源数据进行混合可以实现音色和语言解耦，如果只用 CSMSC 和 LJSpeech 确实音色无法解耦，我们也实验过在 CSMSC + LJSpeech + 1000 条的同一个说话人中英文混合语料训练，也能达到解耦，得到的结论是：
1. CSMSC + LJSpeech 训练（但是我们不是合成单发音人，还是用 2 个 spk_id），中文音素只见过 CSMSC 的，英文音素只见过 LJSpeech 的所以无法解耦
2. 再加一个说话人（这个说话人是中英文混合的，单语的没试过），就能解耦
3. CSMSC + LJSpeech + AISHELL3 + VCTK 可以解耦（我们目前开源的方案）

**Reply by janelu9** [reply]
有开源方言可直接调用的接口吗？

**Reply by lancelee98** [reply]
问题1、CSMSC + LJSpeech 训练（但是我们不是合成单发音人，还是用 2 个 spk_id），无法解耦是什么意思，是音色会串，还是根本训练不出正确的声音呢？ 
问题2、利用当前中英文方案+新的中英文数据进行微调（男声），合成出的声音电流音/机械音很重，您这边有什么建议吗？ mfa也尝试过自己训练，自己的数据干净无噪咬字清晰。

**Comment by janelu9** [comment]
有开源方言可直接调用的接口吗？

**Reply by yazone** [reply]
> 有开源方言可直接调用的接口吗？

暂时没看到有免费的

**Reply by WhiteFu** [reply]
大佬你好，上述数据是网上爬虫整理的，还是采购的，方言效果不错，如果是采购的话请问哪里可以获取

**Reply by yazone** [reply]
> 大佬你好，上述数据是网上爬虫整理的，还是采购的，方言效果不错，如果是采购的话请问哪里可以获取

整理的呢，主要验证无英文语料的方言发音人的英文发音音色和韵律是什么效果

**Reply by WhiteFu** [reply]
好的，谢谢哈

**Reply by 523997931** [reply]
大佬，您是只用了中文语料finetune，没有加英文语料训练出的结果嘛？

**Comment by ly1984** [comment]
@yazone 大佬你好，卡通音色和方言大概用了多少数量的数据啊？另外方言这块例如粤语的声母韵母体系以及MFA使用的是普通话的吗？

**Reply by yazone** [reply]
1、几千条，finetune很快，可能不需要太多数据，关键是覆盖到。
2、粤语用普通话标记有做过尝试，对不懂粤语的人听起来像那么回事，找了广州本地人说每个字像没读全。还有前端需要处理，比如粤语里不说”我是“这个”是“用"系”，另外“我与春风皆过客”，“皆”字粤语里不读"jie"读”gai"吧。

下面是之前尝试的粤语，仔细听听：

https://user-images.githubusercontent.com/10195479/199667927-1aa965fd-bb6f-499b-a2d8-bcb92816b153.mp4

**Reply by kFoodie** [reply]
> 1、几千条，finetune很快，可能不需要太多数据，关键是覆盖到。 2、粤语用普通话标记有做过尝试，对不懂粤语的人听起来像那么回事，找了广州本地人说每个字像没读全。还有前端需要处理，比如粤语里不说”我是“这个”是“用"系”，另外“我与春风皆过客”，“皆”字粤语里不读"jie"读”gai"吧。
> 
> 下面是之前尝试的粤语，仔细听听：
> 
>  tmp_1663293494.6977215.mp4


用粤语finetune，使用的数据是啥样的啊？我用粤语的音频，普通话的文本。然后测试结果出来全是普通话……请问您还做了哪些修改啊

**Reply by yygg678** [reply]
@yazone 请问下卡通音色的数据是爬取的吗？大概多少小时多少条数据？卡因音色的模型有分享吗？谢谢！

**Reply by 13604099691** [reply]
>  tmp_1663293494.6977215.mp4

大佬，请问你finetune用的是aishell3/tts3还是aishell3/ernie_sat

**Reply by kslz** [reply]
> 大佬，请问你finetune用的是aishell3/tts3还是aishell3/ernie_sat

应该是examples/other/tts_finetune/tts3/

**Reply by 13604099691** [reply]
> 应该是examples/other/tts_finetune/tts3/

这个流程我跑了，batch 256，用了1000条数据，数据格式是仿照csmsc，数据处理就是用的你那个小软件，啪一下很快，预训练模型用的fastspeech2_mix_ckpt_1.2.0 和 hifigan_aishell3_ckpt_0.2.0，没有冻结任何层，finetune出来的效果鼻音有点重，试了好几种spkid，并没有你们的效果好，大概率是啥原因嘞

**Reply by kslz** [reply]
> 这个流程我跑了，batch 256，用了1000条数据，数据格式是仿照csmsc，数据处理就是用的你那个小软件，啪一下很快，预训练模型用的fastspeech2_mix_ckpt_1.2.0 和 hifigan_aishell3_ckpt_0.2.0，没有冻结任何层，finetune出来的效果鼻音有点重，试了好几种spkid，并没有你们的效果好，大概率是啥原因嘞

只能猜测是数据集的问题了，另外标注目前有了个自动标注的方案，只需要切片音频，然后会利用asr标注
https://aistudio.baidu.com/aistudio/projectdetail/5003396?channelType=0&channel=0

**Reply by 13604099691** [reply]
> 只能猜测是数据集的问题了，另外标注目前有了个自动标注的方案，只需要切片音频，然后会利用asr标注 https://aistudio.baidu.com/aistudio/projectdetail/5003396?channelType=0&channel=0

标注字幕这块我是用的剪映，字幕生成的效果不错，然后利用你那个小软件切片音频。对了，speajer_id_map.txt文件里的ID是不能新增的吗，finetune的时候只能选择里面的spkid吗

**Reply by kslz** [reply]
> 标注字幕这块我是用的剪映，字幕生成的效果不错，然后利用你那个小软件切片音频。对了，speajer_id_map.txt文件里的ID是不能新增的吗，finetune的时候只能选择里面的spkid吗

这可能得问子龙了@yt605155624

**Reply by yt605155624** [reply]
是的，可微调 speaker 数目的上限是预训练模型的 speaker 数目

**Reply by 13604099691** [reply]
> 是的，可微调 speaker 数目的上限是预训练模型的 speaker 数目

请问，原始音频的采样率对训练结果有影响吗？我看AISHELL-3数据集的音频采样率都是44100，需要保持一致吗

**Reply by yt605155624** [reply]
有一定影响，AISHELL-3数据集的音频采样率都是44100，但是我们的 config 设置的是 24k 哈，建议你的采样率 >=24k，不需要自己执行下采样，我们在 librosa.load 的时候会按照 config 里面设置的采样率进行下载采样

**Reply by maize-j** [reply]
> 有一定影响，AISHELL-3数据集的音频采样率都是44100，但是我们的 config 设置的是 24k 哈，建议你的采样率 >=24k，不需要自己执行下采样，我们在 librosa.load 的时候会按照 config 里面设置的采样率进行下载采样

请问除了采样率以外，还有音频的什么其他因素会对训练结果有影响吗？

**Comment by kslz** [comment]
微调后的样例：
珈乐：（600句）

https://user-images.githubusercontent.com/54951765/202375284-c24cd10c-3f9d-4cb4-be0e-c344cb0ad8f2.mp4

星瞳：（850句）

https://user-images.githubusercontent.com/54951765/202375353-9be6d956-4569-43cc-b7db-c22c2e30f1fc.mp4

唐国强：（45句）

https://user-images.githubusercontent.com/54951765/202375896-1c555a8a-5da4-48f8-976d-72d9ae160da4.mp4

**Reply by 13604099691** [reply]
大佬，再请教一下呀，你这个微调是走的/other/tts_finetune/tts3/   还是/aishell3/tts3/    这两个区别大吗

**Reply by kslz** [reply]
> 大佬，再请教一下呀，你这个微调是走的/other/tts_finetune/tts3/ 还是/aishell3/tts3/ 这两个区别大吗

没有区别，都一样的，只是aishell3/tts3是最初的流程，需要手动复制预训练模型，然后修改jsonl文件之类的，other/tts_finetune/tts3是自动化了一些流程

**Reply by 13604099691** [reply]
> 没有区别，都一样的，只是aishell3/tts3是最初的流程，需要手动复制预训练模型，然后修改jsonl文件之类的，other/tts_finetune/tts3是自动化了一些流程

other/tts_finetune/tts3  这里是连mfa都自动化了对吧，那是不是也就不需要再走一遍other/mfa流程了呢！只是这个一直finetune不出好的结果来，求思路！！！

**Reply by kslz** [reply]
> other/tts_finetune/tts3 这里是连mfa都自动化了对吧，那是不是也就不需要再走一遍other/mfa流程了呢！只是这个一直finetune不出好的结果来，求思路！！！

具体流程你可以参考一下我上面发的AIstudio项目，里面的代码分布调用了finetune的过程，可能会好理解一点。另外结果不好一般就只能从优化数据集来操作了，比如降噪，去混响，去背景音乐，人声增强，我这里有一个adobe的人声增强网站，你可以看看有没有帮助。https://podcast.adobe.com/enhance 再或者就是试试vits或者sovits了

**Reply by 13604099691** [reply]
> 具体流程你可以参考一下我上面发的AIstudio项目，里面的代码分布调用了finetune的过程，可能会好理解一点。另外结果不好一般就只能从优化数据集来操作了，比如降噪，去混响，去背景音乐，人声增强，我这里有一个adobe的人声增强网站，你可以看看有没有帮助。https://podcast.adobe.com/enhance 再或者就是试试vits或者sovits了

[2023-02-09 15:13:27] [INFO] [fastspeech2_updater.py:236] Evaluate: l1_loss: 0.607731, duration_loss: 0.118917, pitch_loss: 0.331540, energy_loss: 0.330861, loss: 1.389049
[2023-02-09 15:13:28] [INFO] [trainer.py:167]  iter: 99296/300, Rank: 0, l1_loss: 0.401198, duration_loss: 0.018145, pitch_loss: 0.274276, energy_loss: 0.250634, loss: 0.944254, avg_reader_cost: 0.06157 sec, avg_batch_cost: 0.16330 sec, avg_samples: 10, avg_ips: 61.23662 sequences/sec
[2023-02-09 15:13:28] [INFO] [fastspeech2_updater.py:236] Evaluate: l1_loss: 0.608492, duration_loss: 0.119395, pitch_loss: 0.331324, energy_loss: 0.326559, loss: 1.385769
[2023-02-09 15:13:29] [INFO] [trainer.py:167]  iter: 99297/300, Rank: 0, l1_loss: 0.402200, duration_loss: 0.019330, pitch_loss: 0.304988, energy_loss: 0.230109, loss: 0.956628, avg_reader_cost: 0.06630 sec, avg_batch_cost: 0.28681 sec, avg_samples: 10, avg_ips: 34.86656 sequences/sec
[2023-02-09 15:13:29] [INFO] [fastspeech2_updater.py:236] Evaluate: l1_loss: 0.608661, duration_loss: 0.119719, pitch_loss: 0.331491, energy_loss: 0.329876, loss: 1.389748
[2023-02-09 15:13:29] [INFO] [trainer.py:167]  iter: 99298/300, Rank: 0, l1_loss: 0.397949, duration_loss: 0.017186, pitch_loss: 0.297324, energy_loss: 0.332177, loss: 1.044635, avg_reader_cost: 0.06658 sec, avg_batch_cost: 0.18014 sec, avg_samples: 10, avg_ips: 55.51348 sequences/sec
[2023-02-09 15:13:30] [INFO] [fastspeech2_updater.py:236] Evaluate: l1_loss: 0.609406, duration_loss: 0.120017, pitch_loss: 0.331668, energy_loss: 0.333987, loss: 1.395078
[2023-02-09 15:13:30] [INFO] [trainer.py:167]  iter: 99299/300, Rank: 0, l1_loss: 0.402587, duration_loss: 0.017864, pitch_loss: 0.274866, energy_loss: 0.245587, loss: 0.940904, avg_reader_cost: 0.06659 sec, avg_batch_cost: 0.17496 sec, avg_samples: 10, avg_ips: 57.15479 sequences/sec
[2023-02-09 15:13:31] [INFO] [fastspeech2_updater.py:236] Evaluate: l1_loss: 0.609953, duration_loss: 0.120349, pitch_loss: 0.332789, energy_loss: 0.340305, loss: 1.403396
[2023-02-09 15:13:31] [INFO] [trainer.py:167]  iter: 99300/300, Rank: 0, l1_loss: 0.401886, duration_loss: 0.016462, pitch_loss: 0.259123, energy_loss: 0.217671, loss: 0.895141, avg_reader_cost: 0.06636 sec, avg_batch_cost: 0.36397 sec, avg_samples: 10, avg_ips: 27.47508 sequences/sec
[2023-02-09 15:13:32] [INFO] [fastspeech2_updater.py:236] Evaluate: l1_loss: 0.609832, duration_loss: 0.120612, pitch_loss: 0.333745, energy_loss: 0.348830, loss: 1.413019

这是我训练时候的损失下降，是不是没到正常水平

**Reply by kslz** [reply]
> 这是我训练时候的损失下降，是不是没到正常水平

不好说，因为据我所知finetune过程中loss不说明问题，只能隔一段时间听一听效果来主观判断

**Reply by JovenRong** [reply]
@kslz 大佬好，打扰了，我听您训练的声音微调的是非常棒的，我目前，other/tts_finetune/tts3流程训练了男声，先是训练1000多句，然后提升到3000句提升比较明显，但还是机器人很重，3000句效果如下（大概2小时音频量）：

https://user-images.githubusercontent.com/3060496/229413832-2e55f590-a0c3-47a1-8c04-f4d36edff287.mp4
我继续增加到4500句（大概3小时音频量），基本没任何提升了，目前我可以调的只有epoch，200, 400，600都试过，lost基本上到1.4左右就下不去，请问大佬按照您训练的经验，我应该怎么继续提升呢？epoch通常设置多少呢？继续增加数据量吗？希望给点建议，感谢。

**Reply by wutl92** [reply]
> 微调后的样例： 珈乐：（600句）
> 
>  3_2.mp4 
> 星瞳：（850句）
> 
>  xtdcs.mp4 
> 唐国强：（45句）
> 
>  4.mp4

你好，请问想把声音读出来，带点情感不那么生硬，是需要自己训练吗？ 还是可以微调什么参数？

**Reply by zapjone** [reply]
@kslz 大佬，你好，听《珈乐》这个合成的声音很自然。请问下这个数据集是使用的哪个数据集呢？能给个数据集的下载地址吗？谢谢!!!

**Reply by kslz** [reply]
> @kslz 大佬，你好，听《珈乐》这个合成的声音很自然。请问下这个数据集是使用的哪个数据集呢？能给个数据集的下载地址吗？谢谢!!!

应该是截的和离书和之后的那期单人直播，手动切的数据集。工具可以参考我的sound_dataset_tools2项目

**Comment by caixxiong** [comment]
您好，请问有具体的方法（做法）介绍嘛？在音素建模上中英文是怎么处理的呢？

**Reply by yt605155624** [reply]
中英文混合语音合成参考
- https://github.com/PaddlePaddle/PaddleSpeech/tree/develop/examples/zh_en_tts/tts3

小样本 finetune 参考
- https://github.com/PaddlePaddle/PaddleSpeech/discussions/1842
- https://github.com/PaddlePaddle/PaddleSpeech/tree/develop/examples/other/tts_finetune/tts3

可以学习一下我们的教程
- https://aistudio.baidu.com/aistudio/course/introduce/25130


<img width="959" alt="image" src="https://user-images.githubusercontent.com/24568452/218389342-0b4fd310-691e-421c-9dc3-47853c533c83.png">

**Comment by yangqinj** [comment]
想问一下大家的卡通人物、明星和不同方言的数据都是从哪里获取的呢？

**Reply by kslz** [reply]
> 想问一下大家的卡通人物、明星和不同方言的数据都是从哪里获取的呢？

来看看这个项目 https://github.com/kslz/sound_dataset_tools2

**Reply by yangqinj** [reply]
感谢回复！不过我的意思不是如何将数据转换为模型需要的格式，而是去哪里获取原始的音频文件。不知道你在哪里收集的？我在网上找到了一些动画片的音频，打算用这些数据构造数据集

**Reply by kslz** [reply]
> 感谢回复！不过我的意思不是如何将数据转换为模型需要的格式，而是去哪里获取原始的音频文件。不知道你在哪里收集的？我在网上找到了一些动画片的音频，打算用这些数据构造数据集

我用的这个软件下载的音频 https://github.com/nilaoda/BBDown

**Reply by yangqinj** [reply]
👍

**Comment by ydmo** [comment]
想问一下，这个方案，看到的栗子都中英混合的，串音确实没有了。但是有试过纯英文，或者纯中文的吗，会串音吗？例如纯英文的，还是用LJSpeech那把声音。

**Comment by kaka1909** [comment]
弱弱的问一句，上面的蜡笔小新的音色，有能使用的finetune后的model么

**Reply by zyb8543d** [reply]
@yt605155624 请问加入多speaker的情感语料(离散情感：开心、伤心、中性、厌恶)可以实现，多语言多speaker的情感解耦吗

**Comment by zyb8543d** [comment]
@yt605155624  请问加入多speaker的情感语料(离散情感：开心、伤心、中性、厌恶)可以实现，多语言多speaker的情感解耦吗

**Comment by xinzhang9525** [comment]
请问下，单发音人方式的中英混合TTS是用的哪个模型呢？也是需要用相关的语料fine-tune吗

**Comment by balicheng** [comment]
基于PaddleSpeech/examples/other/tts_finetune /tts3/的readme，在中英混合模型上，如果从BZNSYP中选出来3k条语音微调am模型，loss可以下降到0.7左右，且用微调模型合成语音，声音比较清晰，同样用aishell3的数据集的某个人的声音的多条数据微调，推理模型合成的声音也很清晰，没有沙沙的声音；
但是用上述方法，在thchs30上选了250个同一个人的语音进行微调，微调后推理模型合成的语音存在沙沙的声音，又从thchs30中选出1000条同一个人的音色的数据微调，微调后loss仍然在1.5左右，且推理合成的声音中存在沙沙的声音，但是能学到微调数据中的音色。
请问大佬们，上述是哪里出现了问题呢

**Reply by starmoon-1134** [reply]
使用am模型推理得到的mel频谱，去微调vocoder模型，应该能缓解这个问题

**Comment by SaltedSlark** [comment]
请问下方言是怎么做到的？

**Comment by jianghuakun** [comment]
大佬们，请问怎么改源码，可以支持中英文混合模型tts streaming?

**Comment by bao2zhao** [comment]
怎么使用这个卡通模型呢？

**Comment by Lingeng56** [comment]
hello，大佬，请问你这个tts是G2P based吗？请问东北话这种方言的g2p要怎么做呢？
